### ISSUE TITLE
[linfa-ica] `GFunc::Exp` computes incorrect g and g' derivatives

### ISSUE THREAD

**Issue by lmmx** [rust-ml/linfa#425]
I think I found a small bug in `fast_ica.rs` (from #47), the `exp` function appears to not actually compute the exponential.

The variable named `exp` stores $-u^2/2$ rather than $exp(-u^2/2)$

For $G(u) = -\exp(-u^2/2)$, the derivatives should be:

$$g(u) = G'(u) = u \exp\left(-\frac{u^2}{2}\right)$$

$$g'(u) = G''(u) = (1 - u^2) \exp\left(-\frac{u^2}{2}\right)$$

https://github.com/rust-ml/linfa/blob/db3cadee926e546a2809346c73689ce5048a96e9/algorithms/linfa-ica/src/fast_ica.rs#L239-L246

Line 240 should be instead:

```rust
    let exp = x.mapv(|x| (-x.powi(2) / A::cast(2.)).exp());
```

- See [scikit-learn](https://github.com/scikit-learn/scikit-learn/blob/8a98df038ce00a7c016f6ce4abdee1f53e81b811/sklearn/decomposition/_fastica.py#L157-L161) for example

```py
def _exp(x, fun_args):
    exp = np.exp(-(x**2) / 2)
    gx = x * exp
    g_x = (1 - x**2) * exp
    return gx, g_x.mean(axis=-1)
```

This means the function returns:

$$g(u) = u \cdot \left(-\frac{u^2}{2}\right) = -\frac{u^3}{2}$$

instead of the intended:

$$g(u) = u \cdot \exp\left(-\frac{u^2}{2}\right)$$

Since $-u^3/2$ is just a constant multiple of $u^3$, and the symmetric decorrelation step normalizes $W$, this makes `GFunc::Exp` behave identically to `GFunc::Cube`.

The proper `exp` contrast downweights large values (because $\exp(-u^2/2) \to 0$ as $|u| \to \infty$), making it robust to outliers while the `cube` contrast does the opposite, large values dominate because $u^3$ grows without bound.
